StardewArchipelago/Items/Unlocks/MagicUnlockManager.cs

```diff
@@ -0,0 +1,284 @@
+        public const string SPELL_SHOCKWAVE_AP = "Spell: Shockwave";
+        public const string SPELL_FIREBALL_AP = "Spell: Fireball";
+        public const string SPELL_FROSTBITE_AP = "Spell: Frostbite";
+        public const string SPELL_TELEPORT_AP = "Spell: Teleport";
+        public const string SPELL_LANTERN_AP = "Spell: Lantern";
+        public const string SPELL_TENDRILS_AP = "Spell: Tendrils";
+        public const string SPELL_PHOTOSYNTHESIS_AP = "Spell: Photosynthesis";
+        public const string SPELL_DESCEND_AP = "Spell: Descend";
+        public const string SPELL_METEOR_AP = "Spell: Meteor";
+        public const string SPELL_BLOODMANA_AP = "Spell: Bloodmana";
+        public const string SPELL_LUCKSTEAL_AP = "Spell: Lucksteal";
+        public const string SPELL_SPIRIT_AP = "Spell: Spirit";
+        public const string SPELL_REWIND_AP = "Spell: Rewind";
+
+        private Dictionary<string, Func<ReceivedItem, LetterAttachment>> _unlockables;
+
+        public MagicUnlockManager()
+        {
+            _unlockables = new Dictionary<string, Func<ReceivedItem, LetterAttachment>>();
+            _unlockables.Add(SPELL_CLEARDEBRIS_AP, SendClearDebrisSpell);
+            _unlockables.Add(SPELL_TILL_AP, SendTillSpell);
+            _unlockables.Add(SPELL_WATER_AP, SendWaterSpell);
+            _unlockables.Add(SPELL_BLINK_AP, SendBlinkSpell);
+            _unlockables.Add(SPELL_EVAC_AP, SendEvacSpell);
+            _unlockables.Add(SPELL_HEAL_AP, SendHealSpell);
+            _unlockables.Add(SPELL_HASTE_AP, SendHasteSpell);
+            _unlockables.Add(SPELL_BUFF_AP, SendBuffSpell);
+            _unlockables.Add(SPELL_DESCEND_AP, SendDescendSpell);
+            _unlockables.Add(SPELL_FROSTBITE_AP, SendFrostbiteSpell);
+            _unlockables.Add(SPELL_FIREBALL_AP, SendFireballSpell);
+            _unlockables.Add(SPELL_TELEPORT_AP, SendTeleportSpell);
+            _unlockables.Add(SPELL_SHOCKWAVE_AP, SendShockwaveSpell);
+            _unlockables.Add(SPELL_TENDRILS_AP, SendTendrilsSpell);
+            _unlockables.Add(SPELL_LANTERN_AP, SendLanternSpell);
+            _unlockables.Add(SPELL_PHOTOSYNTHESIS_AP, SendPhotosynthesisSpell);
+            _unlockables.Add(SPELL_METEOR_AP, SendMeteorSpell);
+            _unlockables.Add(SPELL_BLOODMANA_AP, SendBloodmanaSpell);
+            _unlockables.Add(SPELL_LUCKSTEAL_AP, SendLuckstealSpell);
+            _unlockables.Add(SPELL_SPIRIT_AP, SendSpiritSpell);
+            _unlockables.Add(SPELL_REWIND_AP, SendRewindSpell);
+        }
+
+        public bool IsUnlock(string unlockName)
+        {
+            return _unlockables.ContainsKey(unlockName);
+        }
+
+        public LetterAttachment PerformUnlockAsLetter(ReceivedItem unlock)
+        {
+            return _unlockables[unlock.ItemName](unlock);
+        }
+
+        private LetterAttachment SendClearDebrisSpell(ReceivedItem receivedItem)
+        {
+            foreach (var farmer in Game1.getAllFarmers())
```

Comment:
Why loop over the farmers in every one of these methods and then not care about which farmer it is?